app/src/main/java/camp/nextstep/DispatcherServlet.java

```diff
@@ -8,38 +13,74 @@
+    private final List<HandlerMapping> handlerMappings;
 
     public DispatcherServlet() {
+        handlerMappings = new ArrayList<>();
     }
 
     @Override
     public void init() {
-        this.manualHandlerMapping = new ManualHandlerMapping();
-        this.manualHandlerMapping.initialize();
+        ManualHandlerMapping manualHandlerMapping = new ManualHandlerMapping();
+        manualHandlerMapping.initialize();
+        handlerMappings.add(manualHandlerMapping);
+
+        AnnotationHandlerMapping annotationHandlerMapping = new AnnotationHandlerMapping("camp.nextstep.controller");
+        annotationHandlerMapping.initialize();
+        handlerMappings.add(annotationHandlerMapping);
     }
 
     @Override
     protected void service(final HttpServletRequest request, final HttpServletResponse response) throws ServletException {
         final String requestURI = request.getRequestURI();
         log.debug("Method : {}, Request URI : {}", request.getMethod(), requestURI);
 
+        Object handler = getHandler(request);
+
+        ModelAndView modelAndView = getHandlerAdapter(handler, request, response);
+
         try {
-            final var controller = manualHandlerMapping.getHandler(requestURI);
-            final var viewName = controller.execute(request, response);
-
-            JspView jspView = new JspView(viewName);
-            jspView.render(Collections.emptyMap(), request, response);
-        } catch (Throwable e) {
-            log.error("Exception : {}", e.getMessage(), e);
-            throw new ServletException(e.getMessage());
+            render(modelAndView, request, response);
+        } catch (Exception e) {
+            throw new ServletException("응답을 그리던 중에 예외가 발생했습니다.", e);
         }
     }
+
+    private Object getHandler(HttpServletRequest request) {
+        for (HandlerMapping handlerMapping : handlerMappings) {
+            Object handler = handlerMapping.getHandler(request);
+            if (handler != null) return handler;
+        }
+        return null;
+    }
+
+    private ModelAndView getHandlerAdapter(Object handler, HttpServletRequest request, HttpServletResponse response) {
+        try {
+            if (handler instanceof Controller controller) {
+                JspView view = new JspView(controller.execute(request, response));
+                return new ModelAndView(view);
+            }
+            if (handler instanceof HandlerExecution handlerExecution) {
+                return handlerExecution.handle(request, response);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("핸들러 동작 중에 예외가 발생했습니다.", e);
+        }
+
+        throw new RuntimeException("매핑이 없어요: " + request.getRequestURI());
+    }
+
+    private void render(ModelAndView mav, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        Map<String, Object> model = mav.getModel();
+        mav.getView().render(model, request, response);
```

Comment:
값을 꺼내 직접 연산하기보다는 객체에게 적절한 요청과 응답을 받는걸 권장드려요. 
```suggestion
        Map<String, Object> model = mav.getModel();
        mav.render(request, response);
```